cedar-policy-validator/src/lib.rs

```diff
@@ -206,6 +286,192 @@ impl Validator {
+                    .map(|l| self.check_entity_deref_level_helper(l, max_allowed_level, policy_id))
+                    .collect();
+                Self::min(v)
+            }
+            ExprKind::GetAttr { expr, .. } | ExprKind::HasAttr { expr, .. } => match expr
+                .as_ref()
+                .data()
+            {
+                Some(ty) => {
+                    let child_level_info =
+                        self.check_entity_deref_level_helper(expr, max_allowed_level, policy_id);
+                    match ty {
+                        Type::EntityOrRecord(EntityRecordKind::Entity { .. })
+                        | Type::EntityOrRecord(EntityRecordKind::ActionEntity { .. }) => {
+                            let child_level = child_level_info.0;
+                            let new_level = child_level.decrement();
+                            if new_level.level < Some(0) {
+                                (
+                                    new_level,
+                                    Some(EntityDerefLevelViolation {
+                                        source_loc: e.source_loc().cloned(),
+                                        policy_id: policy_id.clone(),
+                                        actual_level: new_level,
+                                        allowed_level: max_allowed_level.clone(),
+                                    }),
+                                )
+                            } else {
+                                (new_level, None)
+                            }
+                        }
+                        Type::EntityOrRecord(EntityRecordKind::AnyEntity) => {
+                            // AnyEntity cannot be dereferenced
+                            (EntityDerefLevel { level: Some(0) }, None)
+                        }
+                        _ => child_level_info,
+                    }
+                }
+                // PANIC SAFETY: Strict validation passed, so annotating the AST will succeed
+                #[allow(clippy::unreachable)]
+                None => unreachable!("Expected type-annotated AST"),
+            },
+            ExprKind::Like { expr, .. } | ExprKind::Is { expr, .. } => {
+                self.check_entity_deref_level_helper(expr, max_allowed_level, policy_id)
+            }
+            ExprKind::Set(elems) => {
+                let v: Vec<(EntityDerefLevel, Option<_>)> = elems
+                    .iter()
+                    .map(|l| self.check_entity_deref_level_helper(l, max_allowed_level, policy_id))
+                    .collect();
+                Self::min(v)
+            }
+            ExprKind::Record(fields) => {
+                let v: Vec<(EntityDerefLevel, Option<_>)> = fields
+                    .iter()
+                    .map(|(_, l)| {
+                        self.check_entity_deref_level_helper(l, max_allowed_level, policy_id)
+                    })
+                    .collect();
+                Self::min(v)
```

Comment:
This case isn't obvious to me. What's the intended behavior for something like `{foo: principal, bar: <whatever>}.foo`?

Reply:
My understanding of the RFC is that it should recurse into `<whatever>` in determining the maximum number of dereferences that could occur.

Reply:
Actually rereading the RFC, it isn't explicit but "Note that for records, individual attributes may have entities with different levels. E.g., { a: principal, b: principal.manager } would have type { a: User(2), b: User(1) }." sounds like it should only consider `foo` here.

Reply:
That also how I read it

Reply:
Latest revision considers the level of the specific attribute being accessed.